feat: Add Browser MCP for web browsing capabilities

it-baron opened the pull request:
Closes https://github.com/archestra-ai/archestra/issues/1303
/claim https://github.com/archestra-ai/archestra/issues/1303

Summary

- Add BrowserPanel component to display live browser screenshots during Playwright MCP tool execution
- Implement WebSocket-based streaming for real-time screenshot updates from browser automation
- Add browser-stream routes/service with conversation ownership authorization
- Tab session isolation

it-baron commented:
Rebased onto latest `dev` and resolved merge conflicts:

**Resolved conflicts in:**
- `platform/backend/src/clients/chat-mcp-client.ts` - merged extended context parameters with typed `toolArguments`
- `platform/frontend/src/app/chat/page.tsx` - merged Browser panel (this PR) with Artifact panel (from dev)

**Additional fix:**
- Removed duplicate `conversationId` parameters introduced during rebase

All browser feature tests pass (18/18).

brojd commented on platform/backend/src/clients/chat-mcp-client.ts:
```diff
@@ -19,6 +19,31 @@ import {
+    _browserStreamFeature = browserStreamFeature;
+  }
+  return _browserStreamFeature;
+}
```
Can we simplify all this by removing lines 29-45 and then in line 575 in this file:
```
// Lazily loaded browserStreamFeature to avoid circular dependency
const { browserStreamFeature } = await import(
      "@/services/browser-stream-feature"
    );
```
?

brojd commented on platform/backend/src/routes/proxy/utils/adapters/anthropic.ts:
```diff
@@ -213,9 +341,20 @@ export function toolResultsToMessages(
+              hasImages: true,
+            },
+            "Tool result contains images, converting to Anthropic image blocks",
+          );
```
I have few questions around it:
1. Can you explain a bit this part? Is it about the scenario when the tool call returns base64 image and then we convert it to Anthropic format so that LLM can actually understand what's the content of the image? For example to understand the result of `browser_take_screenshot` tool call?
2. Seems this function is about converting image blocks only. The name of the function should reflect that
3. In case those changes are required, we should introduce it in openai and gemini adapters too
4. I'm trying to test a simple scenario. In chat I ask "Go to https://thumbs.dreamstime.com/b/welcome-6723095.jpg, take screenshot and summarize it for me". Then `microsoft__playwright-mcp__browser_take_screenshot` is called, the result includes base64 data. And then immediately the context limit is exceeded because of this base64 string. It's the same for all the providers. Any thoughts around that?

it-baron replied:
2. **Done.** Renamed `convertMcpContentToAnthropic` to `convertMcpImageBlocksToAnthropic` for clarity.

3. **Done.** Added image block conversion to both OpenAI and Gemini adapters:
   - OpenAI: `{ type: "image_url", image_url: { url: "data:mime;base64,..." } }`
   - Gemini: `{ inlineData: { mimeType, data } }`
   - Added comprehensive tests for all three adapters
   - Documented adapter contract in `adapters/index.ts`

4. **Fixed.** Added image stripping from message history before storing to database. After the LLM has processed the image, the base64 data is replaced with `"[Image data stripped to save context]"`. Image parts are converted to text parts to maintain valid message structure. This prevents context limit issues on subsequent turns.

Commits:
- `feat(adapters): add image block conversion for all LLM providers`
- `fix(chat): strip base64 images from messages before storing`

it-baron commented:
## Additional Browser Preview Fixes

Following up on the earlier discussion, I've made several additional fixes:

### 1. Manual Navigation Context
When user navigates via browser panel address bar, a message is now added to the conversation:
- `[User manually navigated browser to: https://example.com]`
- `[User navigated browser back to previous page]`

This ensures the AI knows about manual navigation and will take a new screenshot when asked about the current page.

### 2. Browser Tab Isolation (Critical Fix)
**Root cause found:** The tool cache key didn't include `conversationId`. When tools were cached, the `conversationId` was baked into closures. When switching conversations, cached tools still used the old conversation's tab!

**Fix:** Added `conversationId` to `getToolCacheKey()` so each conversation gets its own tool cache with correct tab selection.

### 3. Auto-close Browser Panel
- Panel closes when switching to a profile without Playwright tools
- Panel doesn't render if current profile lacks Playwright tools

### 4. UI Improvements
- Zoom icon: Shows `ZoomIn` when zoomed out, `ZoomOut` at 100%
- Removed decorative Move/Globe icons (every icon now has an action)
- Fixed minimized view extra line (removed border-b when minimized)

Commit: `fix(browser-stream): multiple browser preview improvements`

brojd commented on platform/backend/src/routes/browser-stream.ts:
```diff
@@ -0,0 +1,317 @@
+        request.organizationId,
+      );
+      if (!agentId) {
+        return reply.send({
```
Question about all the failure cases in this file, is there any specific reason to not `throw new ApiError()` which is then handled by https://github.com/archestra-ai/archestra/blob/dev/platform/backend/src/server.ts#L176-L221
?

it-baron replied:
Good catch! Refactored all routes to use `throw new ApiError()` - removed manual error responses and `error` fields from response schemas. Errors are now handled by the global Fastify error handler.

brojd commented on platform/backend/src/routes/proxy/adapterV2/mcp-image.ts:
```diff
@@ -0,0 +1,109 @@
+export type McpImageBlock = {
```
I suggest to move this file (and corresponding test file) to proxy/utils folder

brojd commented on platform/frontend/src/lib/chat.query.ts:
```diff
@@ -295,3 +295,17 @@ export function usePromptTools(promptId: string | undefined) {
+      const toolName = tool.name;
+      return (
+        typeof toolName === "string" &&
+        (toolName.includes("playwright") || toolName.startsWith("browser_"))
```
Can we create some simple util for this check in `shared/utils` to not repeat hardcoded "playwright" value in multiple places?

<img width="767" height="456" alt="Image" src="https://github.com/user-attachments/assets/c02c8f80-426a-4dc3-b0c3-5ed8e585af0b" />

brojd commented:
Sometimes when I close the chats, browser tabs are not cleaned
<img width="1223" height="1030" alt="image" src="https://github.com/user-attachments/assets/af904685-9067-47f8-a2f9-5743c0130834" />

it-baron commented:
## Updates from review feedback

Thanks for the review! I've addressed the feedback:

### 1. Moved `mcp-image.ts` to `proxy/utils/` folder
- Moved `adapterV2/mcp-image.ts` → `utils/mcp-image.ts`
- Updated imports in adapterV2 adapters
- Made `utils/adapters/mcp-image.ts` re-export from parent for backwards compatibility

### 2. Created `isBrowserMcpTool` utility in `shared/utils.ts`
- Extracted the `toolName.includes("playwright") || toolName.startsWith("browser_")` pattern
- Updated all usages in:
  - `frontend/src/lib/chat.query.ts`
  - `backend/src/clients/chat-mcp-client.ts`
  - `backend/src/services/browser-stream.ts`

### 3. Refactored error handling to use `ApiError`
- Replaced manual error responses with `throw new ApiError()` in browser-stream routes
- Removed `error` field from response schemas (errors now handled by global handler)

### 4. Fixed browser tab cleanup issue
- `closeTab` was returning early when tab index wasn't in memory (after server restart)
- Now lists browser tabs and closes the most recent one as best-effort cleanup

brojd commented on platform/backend/src/routes/proxy/openai.ts:
```diff
@@ -395,6 +401,30 @@ const openAiProxyRoutes: FastifyPluginAsyncZod = async (fastify) => {
         toonCostSavings = stats.toonCostSavings;
       }
 
+      // Strip large browser tool results to reduce token usage
```
ah nvm

it-baron commented:
@brojd addressed the merge checklist:

- Restored try/catch around MCP tool execution in `platform/backend/src/clients/chat-mcp-client.ts`.
- Reverted changes in legacy adapters (`platform/backend/src/routes/proxy/utils/adapters/*`) to match dev and minimize diff.
- Removed new behavior from v1 proxy routes and implemented provider-specific logic in v2 adapters (`platform/backend/src/routes/proxy/adapterV2/openai.ts`, `platform/backend/src/routes/proxy/adapterV2/anthropic.ts`, `platform/backend/src/routes/proxy/adapterV2/gemini.ts`). The only remaining v1 change is a type-only cast in `platform/backend/src/routes/proxy/openai.ts` for SDK compatibility (no runtime impact).
- Gated all backend behavior changes (tool-result stripping, MCP image conversion, message stripping in `platform/backend/src/routes/chat/routes.ts`) behind existing `FEATURES_BROWSER_STREAMING_ENABLED`

Ran `pnpm -C platform run check:ci`: backend is clean; remaining failures are unrelated frontend Zod resolver version mismatches (per request to ignore).

it-baron commented:
@brojd rebased on origin/dev and resolved the merge conflict in `platform/frontend/src/lib/chat.query.ts` (merged `SupportedProvider` + `isBrowserMcpTool` import). 

Items are covered: try/catch restored in `chat-mcp-client.ts`, legacy adapters diff minimized, 
v1 proxy changes removed in favor of adapterV2/llm-proxy-handler, 
and backend behavior gated behind `FEATURES_BROWSER_STREAMING_ENABLED`.

Additional fixes since your note:
- `browser_tabs` JSON current URL parsing (with test) so address bar sync works when tabs return JSON.
- MCP image size guard in adapterV2 (OpenAI/Anthropic/Gemini) with placeholder, behind the same feature flag.
- MCP HTTP concurrency limiter (flag-gated) and tab selection robustness already included.

Checks: `pnpm -C platform/backend run check:ci` passed. 
Knip: config hints only (redundant entry patterns).
